File: backend/app/models/compound_event.py
from datetime import datetime
from typing: List, Optional, Dict, Any
from beanie import Document, Indexed
from pydantic import Field
from pymongo import ASCENDING, DESCENDING, GEOSPHERE


class CompoundEvent(Document):
    start_time: Indexed(datetime)
    end_time: Indexed(datetime)
    detected_at: Indexed(datetime) = Field(default_factory=datetime.utcnow)
    expires_at: Indexed(datetime)
    centroid: Dict[str, Any]
    radius_km: float
    domains: List[str]
    event_ids: List[str]
    event_count: int
    time_span_hours: float
    severity: Indexed(float)
    severity_tier: str
    amplification_factor: float = 1.0
    news_headlines: List[str] = Field(default_factory=list)
    status: str = "active"
    metadata: Dict[str, Any] = Field(default_factory=dict)
    
    class Settings:
        name = "compound_events"
        indexes = [
            [("start_time", DESCENDING)],
            [("detected_at", DESCENDING)],
            [("severity", DESCENDING), ("detected_at", DESCENDING)],
            [("status", ASCENDING)],
            [("centroid", GEOSPHERE)],
            [("domains", ASCENDING)],
            [("expires_at", ASCENDING)],
        ]

File: frontend/src/types/CompoundEvent.ts
export interface CompoundEvent {
  id: string
  detected_at: string
  expires_at: string
  centroid: {
    type: 'Point'
    coordinates: [number, number]
  }
  radius_km: number
  domains: string[]
  event_ids: string[]
  severity: number
  severity_tier: 'info' | 'low' | 'moderate' | 'high' | 'critical'
  news_headlines: string[]
  status: 'active' | 'expired' | 'acknowledged'
  metadata: Record<string, any>
}

export interface CompoundEventListResponse {
  events: CompoundEvent[]
  total: number
  page: number
  page_size: number
}

export interface StatCardsResponse {
  active_high_severity_clusters: number
  total_events_last_hour: number
  most_active_domain: string
}